File: mobile/ios/PrestamosApp/Utils/CedulaValidator.swift
import Foundation

struct CedulaValidator {
    
    /// Valida formato de cédula dominicana
    static func validarFormato(_ cedula: String) -> Bool {
        let cleaned = cedula.replacingOccurrences(of: "[\\s-]", with: "", options: .regularExpression)
        
        guard cleaned.count == 11 else { return false }
        guard cleaned.allSatisfy({ $0.isNumber }) else { return false }
        
        return true
    }
    
    /// Valida cédula usando algoritmo de verificación
    static func validarCedula(_ cedula: String) -> Bool {
        guard validarFormato(cedula) else { return false }
        
        let cleaned = cedula.replacingOccurrences(of: "[\\s-]", with: "", options: .regularExpression)
        let multiplicadores = [1, 2, 1, 2, 1, 2, 1, 2, 1, 2]
        
        var suma = 0
        for i in 0..<10 {
            let digito = Int(String(cleaned[cleaned.index(cleaned.startIndex, offsetBy: i)])) ?? 0
            let producto = digito * multiplicadores[i]
            suma += producto > 9 ? producto - 9 : producto
        }
        
        let digitoVerificador = Int(String(cleaned[cleaned.index(cleaned.startIndex, offsetBy: 10)])) ?? 0
        let residuo = suma % 10
        let verificadorEsperado = residuo == 0 ? 0 : 10 - residuo
        
        return digitoVerificador == verificadorEsperado
    }
    
    /// Formatea cédula con guiones (000-0000000-0)
    static func formatearCedula(_ cedula: String) -> String {
        let cleaned = cedula.replacingOccurrences(of: "[\\s-]", with: "", options: .regularExpression)
        
        guard cleaned.count == 11 else { return cedula }
        
        let index1 = cleaned.index(cleaned.startIndex, offsetBy: 3)
        let index2 = cleaned.index(cleaned.startIndex, offsetBy: 10)
        
        return "\(cleaned[..<index1])-\(cleaned[index1..<index2])-\(cleaned[index2...])"
    }
}

File: mobile/android/app/src/main/java/com/erp/prestamos/utils/CedulaValidator.java
package com.erp.prestamos.utils;

public class CedulaValidator {
    
    /**
     * Valida formato de cédula dominicana
     * @param cedula Cédula a validar
     * @return true si el formato es válido
     */
    public static boolean validarFormato(String cedula) {
        if (cedula == null || cedula.isEmpty()) {
            return false;
        }
        
        // Remover guiones y espacios
        cedula = cedula.replaceAll("[\\s-]", "");
        
        // Debe tener 11 dígitos
        if (cedula.length() != 11) {
            return false;
        }
        
        // Debe contener solo números
        if (!cedula.matches("\\d+")) {
            return false;
        }
        
        return true;
    }
    
    /**
     * Valida cédula usando algoritmo de verificación
     * @param cedula Cédula a validar
     * @return true si la cédula es válida
     */
    public static boolean validarCedula(String cedula) {
        if (!validarFormato(cedula)) {
            return false;
        }
        
        cedula = cedula.replaceAll("[\\s-]", "");
        
        // Algoritmo de validación de cédula dominicana
        int suma = 0;
        int[] multiplicadores = {1, 2, 1, 2, 1, 2, 1, 2, 1, 2};
        
        for (int i = 0; i < 10; i++) {
            int digito = Character.getNumericValue(cedula.charAt(i));
            int producto = digito * multiplicadores[i];
            suma += (producto > 9) ? producto - 9 : producto;
        }
        
        int digitoVerificador = Character.getNumericValue(cedula.charAt(10));
        int residuo = suma % 10;
        int verificadorEsperado = (residuo == 0) ? 0 : 10 - residuo;
        
        return digitoVerificador == verificadorEsperado;
    }
    
    /**
     * Formatea cédula con guiones (000-0000000-0)
     * @param cedula Cédula sin formato
     * @return Cédula formateada
     */
    public static String formatearCedula(String cedula) {
        if (cedula == null || cedula.isEmpty()) {
            return "";
        }
        
        cedula = cedula.replaceAll("[\\s-]", "");
        
        if (cedula.length() == 11) {
            return cedula.substring(0, 3) + "-" + 
                   cedula.substring(3, 10) + "-" + 
                   cedula.substring(10);
        }
        
        return cedula;
    }
}

